Parsing new expression uses incorrect variable when multiple are available

```
======================================== FAILURES =========================================
_______________________________ test_new_expression_parsing _______________________________

    def test_new_expression_parsing():
        @proc
        def bar(n: size, arr: R[n] @ DRAM):
            for i in seq(0, n):
                tmp: R
                tmp = 1.0
                arr[i] = tmp
            i: R
            i = 1.0

>       expand_dim(
            bar, "tmp : _", "n", "i"
        )

tests/test_schedules.py:1604:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/exo/API_scheduling.py:101: in __call__
    return self.func(*bound_args.args, **bound_args.kwargs)
src/exo/API_scheduling.py:1147: in expand_dim
    CheckEffects(new_proc_c._loopir_proc)
src/exo/effectcheck.py:537: in __init__
    body_eff = self.map_stmts(self.orig_proc.body)
src/exo/effectcheck.py:1016: in map_stmts
    sub_body_eff = self.map_stmts(stmt.body)
src/exo/effectcheck.py:1044: in map_stmts
    self.check_bounds(stmt.name, shape, body_eff)
src/exo/effectcheck.py:791: in check_bounds
    self.check_in_bounds(sym, shape, e, y)
src/exo/effectcheck.py:775: in check_in_bounds
    e = self.expr_to_smt(e)
src/exo/effectcheck.py:612: in expr_to_smt
    return self.sym_to_smt(expr.name, expr.type)
src/exo/effectcheck.py:587: in sym_to_smt
    return self.env[sym]
/usr/lib/python3.9/collections/__init__.py:941: in __getitem__
    return self.__missing__(key)            # support subclasses that define __missing__
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = ChainMap({}, {i_3: i_3}, {n_1: n_1, arr_stride_0_7: arr_stride_0_7, tmp_stride_0_8: tmp_stride_0_8}, {})
key = i_5

    def __missing__(self, key):
>       raise KeyError(key)
E       KeyError: i_5

/usr/lib/python3.9/collections/__init__.py:933: KeyError
================================= short test summary info =================================
FAILED tests/test_schedules.py::test_new_expression_parsing - KeyError: i_5
==================================== 1 failed in 0.19s ====================================
```

During parsing, variable `i_5` is captured as the variable `i` from the new expression. However, this is the variable `i` outside the loop definition.

--- a/src/exo/parse_fragment.py
+++ b/src/exo/parse_fragment.py
@@ -52,6 +52,7 @@ class BuildEnv(LoopIR_Do):
         self.result = None
         self.trg = stmt
         self.proc = proc
+        self.halt = False
 
         for a in self.proc.args:
             self.env[a.name] = a.type
@@ -72,7 +73,10 @@ class BuildEnv(LoopIR_Do):
 
     def do_s(self, s):
         if s == self.trg:
-            self.result = self.env.copy()
+            self.result = self.env
+            self.halt = True
+        if self.halt:
+            return
 
         styp = type(s)
         if styp is LoopIR.Assign or styp is LoopIR.Reduce:

--- a/tests/golden/test_schedules/test_new_expr_multi_vars.txt
+++ b/tests/golden/test_schedules/test_new_expr_multi_vars.txt
@@ -1,0 +1,7 @@
+def bar(n: size, arr: R[n] @ DRAM):
+    for i in seq(0, n):
+        tmp: R[n] @ DRAM
+        tmp[i] = 1.0
+        arr[i] = tmp[i]
+    i: R @ DRAM
+    i = 1.0

--- a/tests/test_schedules.py
+++ b/tests/test_schedules.py
@@ -1484,3 +1484,17 @@ def test_stage_mem_accum2(golden):
     accum = stage_mem(accum, "for i in _:_", "out[k, 0:16, 0:16]", "o")
 
     assert str(simplify(accum)) == golden
+
+
+def test_new_expr_multi_vars(golden):
+    @proc
+    def bar(n: size, arr: R[n] @ DRAM):
+        for i in seq(0, n):
+            tmp: R @ DRAM
+            tmp = 1.0
+            arr[i] = tmp
+        i: R @ DRAM
+        i = 1.0
+
+    bar = expand_dim(bar, "tmp : _", "n", "i")
+    assert str(bar) == golden

